Source/Core/DolphinQt/GCMemcardManager.cpp

```diff
@@ -293,162 +343,316 @@ void GCMemcardManager::SetSlotFileInteractive(int slot)
+    ModalMessageBox::warning(this, tr("Import Failed"), error_messages.join(QLatin1Char('\n')));
+    return;
+  }
 
-    UpdateSlotTable(i);
+  for (const Memcard::Savefile& savefile : savefiles)
+  {
+    const auto result = card->ImportFile(savefile);
+
+    // we've already checked everything that could realistically fail here, so this should only
+    // happen if the memory card data is corrupted in some way
+    if (result != Memcard::GCMemcardImportFileRetVal::SUCCESS)
+    {
+      const std::string filename = Memcard::GenerateFilename(savefile.dir_entry);
+      ModalMessageBox::warning(
+          this, tr("Import Failed"),
+          tr("Failed to import \"%1\".").arg(QString::fromStdString(filename)));
+      break;
+    }
+  }
+
+  if (!card->Save())
+  {
+    ModalMessageBox::warning(this, tr("Import Failed"),
+                             tr("Failed to write modified memory card to disk."));
   }
+
+  UpdateSlotTable(slot);
 }
 
-void GCMemcardManager::DeleteFiles()
+void GCMemcardManager::ImportFile()
 {
-  auto selection = m_slot_table[m_active_slot]->selectedItems();
-  auto& memcard = m_slot_memcard[m_active_slot];
+  auto& card = m_slot_memcard[m_active_slot];
+  if (!card)
+    return;
 
-  auto count = selection.count() / m_slot_table[m_active_slot]->columnCount();
+  const QStringList paths = QFileDialog::getOpenFileNames(
+      this, tr("Import Save File(s)"), QString::fromStdString(File::GetUserPath(D_GCUSER_IDX)),
+      QStringLiteral("%1 (*.gci *.gcs *.sav);;%2 (*.gci);;%3 (*.gcs);;%4 (*.sav);;%5 (*)")
+          .arg(tr("Supported file formats"), GetFormatDescription(Memcard::SavefileFormat::GCI),
+               GetFormatDescription(Memcard::SavefileFormat::GCS),
+               GetFormatDescription(Memcard::SavefileFormat::SAV), tr("All Files")));
 
-  // Ask for confirmation if we are to delete multiple files
-  if (count > 1)
-  {
-    QString text = count == 1 ? tr("Do you want to delete the selected save file?") :
-                                tr("Do you want to delete the %1 selected save files?").arg(count);
+  if (paths.isEmpty())
+    return;
 
-    auto response = ModalMessageBox::question(this, tr("Question"), text);
-    ;
+  std::vector<Memcard::Savefile> savefiles;
+  savefiles.reserve(paths.size());
 
-    if (response == QMessageBox::Abort)
+  for (const QString& path : paths)
+  {
+    auto read_result = Memcard::ReadSavefile(path.toStdString());
+    auto* savefile = std::get_if<Memcard::Savefile>(&read_result);
+    if (savefile)
+    {
+      savefiles.emplace_back(std::move(*savefile));
+      continue;
+    }
+
+    const auto* error_code = std::get_if<Memcard::ReadSavefileErrorCode>(&read_result);
+    if (error_code)
+    {
+      ModalMessageBox::warning(
+          this, tr("Import Failed"),
+          tr("Failed to import \"%1\".\n%2").arg(path, GetErrorMessageForErrorCode(*error_code)));
       return;
+    }
+
+    // should never get here but just in case...
```

Comment:
I feel like a variant visitor would remove this code section and avoid the if checks above

Reply:
Honestly I'm not sure if that's really much better.

https://gist.github.com/AdmiralCurtiss/a363a3806c49de9c1b8a77e7b2b2cce6

Thoughts?

Reply:
Alright I massaged this a bit more and I think I ended up with something better. Also has the nice side-effect of showing all errors instead of just the first.

![image](https://user-images.githubusercontent.com/4522237/105930693-47992f80-604a-11eb-9da3-72ed68fb2f62.png)

Reply:
Yeah, that looks pretty good.  I do tend to prefer the class version because it allows for state and doesn't require that weird overloaded notation (though can use `constexpr` with a template) but I think it's a little more readable compared to what was there before

Reply:
Whoops, was busy writing a message and edited multiple times, missed your new comment :)

Reply:
LGTM